keps/74-support-argo-workflow/README.md

```diff
@@ -0,0 +1,722 @@
+
+#### How to suspend the whole workflow
+
+Given the diverse resource, node affinity, and toleration requirements among workflow pods, 
+determining the necessary resources for each flavor becomes challenging for the controller. 
+Users must specify workflow resources via annotations like kueue.k8s.io/max-resources, tolerations 
+with kueue.k8s.io/toleration, and node selectors with kueue.k8s.io/node-selector.
+
+In this way, only one workload will be create for one workflow, so workflows can be integrated into 
+Kueue with current job framework. Only need to implement for the specific workflow types.
+
+#### Drawback and Limitations
+
+- Inability to set distinct nodeSelectors and tolerations for multiple pod sets within a workflow.
+- Hard to inject any node selectors that Kueue could calculate from flavor assignment.
+
+#### Advantages
+
+- Simplified architecture facilitating straightforward implementation. Workload managers can 
+follow the same api to queue their workflows.
+
+### Stage as An Unit
+
+#### Workflow Framework in Kueue
+
+When we regard stage as an unit, multiple workloads will be created for one workflow. So the current
+job framework is not suitable for this case. We need to create a new `workflow framework` for the 
+workflows. The generic workflow framework reconcile workflows in following way:
+
+``` go
+type GenericWorkflow interface {
+  GetActiveStages() []GenericJob
+}
+
+
+func (r *workflowReconciler)ReconcileGenericWorkflow(ctx context.Context, req ctrl.Request, wf GenericWorkflow) {
+  dropFinalizers := GetWorkflowInstance()
+  if dropFinalizers {
+    DropFinalizers()
+    return 
+  }
+
+  // currently we do not support parent-workload for workflow
+  // stage should be implemented as GenericJob
+  stages := wf.GetActiveStages()
+  for _, stage := range stages {
+    // 1. make sure there is only a single existing instance of the workload.
+    // If there's no workload exists and stage is unsuspended, we'll stop it immediately.
+    // Stage should support suspend and resume
+
+    // 1.1 If the workload is pending deletion, suspend the stage if needed
+    // and drop the finalizer.
+
+    // 2. handle stage is finished.
+
+    // 3. handle workload is nil.
+
+    // 4. update reclaimable counts if implemented by the stage
+
+    // 5. handle WaitForPodsReady only for a standalone job.
+    // handle a job when waitForPodsReady is enabled, and it is the main job
+
+    // 6. handle eviction
+
+	  // 7. handle job is suspended.
+
+	  // 8. handle job is unsuspended.
+    r.ReconcileStage(ctx, req, stage, wf)
+  }
+}
+```
+
+Each workflow managers can implement their own controller to integrate with the workflow framework.
+
+#### How to suspend by stages
+
+In my daily work, I primarily encounter workflow managers in the form of Argo and Tekton. Consequently, this section will focus on these two. Other workflow management systems can draw upon the discussions in this section for implementation guidance.
```

Comment:
Do we have any significant differences between Argo/Tekton and Airflow when we execute workflows in Kubernetes?
cc @shravan-achar @akshaychitneni @bigsur0

Reply:
It seems suspend action in Tekton is not supported.

Reply:
To answer @andreyvelich question:

### Argo

Argo seems to have generic support for workloads via custom CRD specs. For example (metaflow does this) One can support a jobset as a node in a dag for workflow. Generally the default integration for Argo is pods. For argo, it seems that one can put labels/annotations on a pod template as part of the workflow definition. 

Argo has the concept of workflows and one can define resusable templates (via WorkflowTemplates) but when you create a workflow that usually means that the workflow will be run once it is created.

### Tekton

Tekton seems to only allow the creation of tasks which in turn create pods. One can combine tasks into a pipeline. A pipeline is a reusable component. Tasks and Pipelines are created into the API and they are essentially saved for reusue. A user would create either a TaskRun or a PipelineRun to execute the task or pipeline. A user is able to inject metadata into the pipelineRun (TaskRun) if they add labels to the PipelineRun. This will inject those labels/annotations on each task.

It is also possible to specify different labels/annotations on each task also.

### Airflow

Airflow is a workflow engine where one defines dags as a python file that will call out to a operator to execute that task. AFAIK Airflow provides support for Job, Pod and SparkOperator (see https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html#). I think Airflow support would work in Kueue also via metadata labels.